sql/core_clients_last_seen_raw_v1.sql

```diff
@@ -0,0 +1,86 @@
+  -- Equivalent to, but more efficient than, calling udf_bitmask_range(1, 28)
+CREATE TEMP FUNCTION
+  udf_bitmask_lowest_28() AS (0x0FFFFFFF);
+  --
+CREATE TEMP FUNCTION
+  udf_shift_one_day(x INT64) AS (IFNULL((x << 1) & udf_bitmask_lowest_28(),
+      0));
+  --
+CREATE TEMP FUNCTION
+  udf_combine_days(prev INT64,
+    curr INT64) AS (udf_shift_one_day(prev) + IFNULL(curr,
+      0));
+  --
+WITH
+  _current AS (
+  SELECT
+    -- In this raw table, we capture the history of activity over the past
+    -- 28 days for each usage criterion as a single 64-bit integer. The
+    -- rightmost bit represents whether the user was active in the current day.
+    CAST(TRUE AS INT64) AS days_seen_bits,
+    DATE_DIFF(submission_date, profile_date, DAY) AS days_since_created_profile,
+    CAST(NULL AS BOOLEAN) AS ping_seen_within_6_days_of_profile_creation,
+    * EXCEPT (submission_date)
+  FROM
+    core_clients_daily_v1
+  WHERE
+    submission_date = @submission_date ),
+  --
+  _previous AS (
+  SELECT
+    * EXCEPT (submission_date) REPLACE (
+      -- Scrub values outside 28 day window.
+    IF
+      (days_since_created_profile BETWEEN 0
+        AND 26,
+        days_since_created_profile,
+        NULL) AS days_since_created_profile)
+  FROM
+    core_clients_last_seen_raw_v1 AS cls
+  WHERE
+    submission_date = DATE_SUB(@submission_date, INTERVAL 1 DAY)
+    -- Filter out rows from yesterday that have now fallen outside the 28-day window.
+    AND udf_shift_one_day(days_seen_bits) > 0),
+  --
+  _joined AS (
+  SELECT
+    @submission_date AS submission_date,
+  IF
+    (_current.client_id IS NOT NULL,
+      _current,
+      _previous).* REPLACE ( udf_combine_days(_previous.days_seen_bits,
+        _current.days_seen_bits) AS days_seen_bits,
+      -- We want to base new profile creation date on the first profile_creation_date
+      -- value we observe, so we propagate an existing non-null value in preference
+      -- to a non-null value on today's observation.
+      COALESCE(_previous.days_since_created_profile + 1,
+        _current.days_since_created_profile) AS days_since_created_profile,
+      -- We only trust profile_creation_date if we see a ping within one week,
+      -- so we calculate this on day 6 and propagate to subsequent days.
+    IF
+      (COALESCE(_previous.days_since_created_profile + 1,
+          _current.days_since_created_profile) = 6,
+        TRUE,
+        _previous.ping_seen_within_6_days_of_profile_creation) AS ping_seen_within_6_days_of_profile_creation)
```

Comment:
we could combine these into a `days_since_seen_within_6_days_of_profile_creation_bits` criterion tracker and avoid a lot of complexity here in favor of a bit mask

Reply:
This is a great idea! I have pushed a new commit to implement this and it indeed removes a good chunk of bespoke logic.